### DISCUSSION TITLE
Event Detection in YOLO: Is There a Built-in Mechanism?

### DISCUSSION THREAD

**Discussion by Manishthakur2503** [ultralytics/ultralytics#19872] [General]
Hello Ultralytics Team,

I’m exploring the possibility of event detection using YOLO. While I understand that YOLO is primarily designed for object detection, I would like to know if there is any built-in support for event detection, such as:
- Any markup or annotation format to define events?
- Native support for tracking object interactions over time?
- Recommended approaches for integrating YOLO with event detection logic?

My requirement is to detect events in aircraft management, such as:
- Fueling
- Baggage unloading and loading
- Passenger boarding and deboarding
- Ground crew activities

If there's no direct support, what would be the best way to implement event detection with YOLO? 

Looking forward to your insights. Thanks in advance

Manish Thakur 
manish.thakur@manomayaai.com

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Manishthakur2503, thank you for reaching out and for your interest in Ultralytics 🚀! 

While YOLO models are optimized for object detection tasks, your query about event detection is intriguing. For detailed insights, I recommend checking out our [Docs](https://docs.ultralytics.com/), which provide guidance on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, as well as useful tips for extending YOLO for advanced use cases.

If this is a feature request or a custom application ❓, please provide additional details about your approach or any progress made. For example:
- Dataset examples or annotations used
- Specific configurations or integrations attempted with YOLO
- Logs or outputs, if applicable

If this is a 🐛 Bug Report related to YOLO performance or functionality, kindly provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to assist in debugging.

## Community and Support

For real-time discussions and brainstorming, feel free to join the Ultralytics community:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for active chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) for knowledge-sharing

## Upgrade

Ensure you're using the latest `ultralytics` package to access the most recent features and fixes:

```bash
pip install -U ultralytics
```

## Environments

For testing and experimentation, consider running YOLO in one of these verified environments:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  
This badge indicates the current status of our CI tests, which ensure YOLO's capabilities across macOS, Windows, and Ubuntu.

An Ultralytics engineer will review your query soon to provide further assistance. Meanwhile, we encourage you to explore the resources above to kickstart your implementation. Good luck with your event detection project! 🚀

**Comment by glenn-jocher** [comment]
Thank you for your question! While YOLO models excel at object detection and tracking (via our [Track mode](https://docs.ultralytics.com/modes/track/)), event detection like fueling/baggage activities would require analyzing temporal object interactions using the detected outputs. We recommend:  
1. Use YOLO for object detection (ground crew, baggage carts, etc.)  
2. Track objects with our built-in [BoT-SORT](https://docs.ultralytics.com/modes/track/#available-trackers) or ByteTrack  
3. Implement custom logic (state machines, temporal analysis) on tracked object trajectories/interactions to detect specific events  

The Ultralytics team appreciates your creative application of YOLO in aviation management!

**Reply by Manishthakur2503** [reply]
@glenn-jocher Thank you for your response.
I appreciate the insights on using YOLO for tracking. I have a couple of follow-up questions:
**1. Integrating YOLO’s Tracking Outputs for Event Detection:** Do you have any recommended approaches or external libraries for integrating YOLO’s tracking outputs with state machines or temporal models to enable event detection?
**2. Handling Multiple Object Interactions in Complex Scenarios:** In scenarios where multiple objects interact simultaneously (e.g., multiple baggage carts unloading at different aircraft), what techniques or post-processing steps do you suggest for differentiating distinct events?

Looking forward to your insights. Thanks in advance.

Manish Thakur
manish.thakur@manomayaai.com

**Reply by glenn-jocher** [reply]
@Manishthakur2503 thank you for your follow-up questions. For event detection integration:  
1. Consider pairing YOLO tracking outputs with temporal logic libraries like Python's [transitions](https://github.com/pytransitions/transitions) for state machines, or temporal modeling with PyTorch  
2. For complex interactions, implement spatial-temporal analysis using unique object IDs from tracking outputs combined with OpenCV positional data and Scikit-learn clustering  

The Ultralytics team appreciates your innovative application of YOLO in aviation logistics! 🚀

**Reply by Manishthakur2503** [reply]
@glenn-jocher ,Thank you for your previous insights.

I have implemented an event detection system using YOLOv8 for object detection and DeepSORT for tracking, combined with spatial-temporal analysis for event recognition. The system:
- Uses DeepSORT to track objects over time.
- Detects events based on object proximity and movement (e.g., aircraft stopping at the gate, baggage handling, refueling).
- Implements custom logic for event detection, leveraging positional changes and unique object IDs.

For reference , this is the code:
```from ultralytics import YOLO
import cv2
import numpy as np
from deep_sort_realtime.deepsort_tracker import DeepSort

# Load YOLOv8 model
model = YOLO("best.pt")

# Initialize DeepSORT tracker
tracker = DeepSort(max_age=30)

# Open video file
cap = cv2.VideoCapture("Ryanair turnaround - Bratislava.mkv")
frame_width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
frame_height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))

# Store object states
aircraft_positions = {}
tracked_objects_dict = {}
event_states = {}  # Prevent duplicate event logs

def is_near(obj1, obj2, threshold=50):
    """Check if two objects are near each other based on bounding box centers."""
    x1_min, y1_min, x1_max, y1_max = obj1
    x2_min, y2_min, x2_max, y2_max = obj2

    center1 = (round((x1_min + x1_max) / 2), round((y1_min + y1_max) / 2))
    center2 = (round((x2_min + x2_max) / 2), round((y2_min + y2_max) / 2))

    return np.linalg.norm(np.array(center1) - np.array(center2)) < threshold

def detect_aircraft_events(track_id, movement):
    """Handle aircraft-specific event detection."""
    if movement < 5 and not event_states.get(f"stopped_{track_id}", False):
        print("EVENT: Aircraft Stopped at Gate")
        event_states[f"stopped_{track_id}"] = True
    
    if movement > 10 and event_states.get(f"stopped_{track_id}", False):
        print("EVENT: Pushback Started")
        event_states[f"stopped_{track_id}"] = False

def detect_proximity_events(track_id, bbox, current_objects, proximity_classes, event_name, threshold=50):
    """Generic proximity event detection."""
    for target_class in proximity_classes:
        for obj in current_objects.get(target_class, []):
            if is_near(bbox, obj.to_tlbr(), threshold):
                if not event_states.get(f"{event_name}_{track_id}", False):
                    print(f"EVENT: {event_name}")
                    event_states[f"{event_name}_{track_id}"] = True
                return
    event_states[f"{event_name}_{track_id}"] = False

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break
    
    # Detect objects with YOLOv8
    results = model(frame, conf=0.5)
    detections = []
    
    if results and results[0].boxes:
        for r in results:
            for box in r.boxes:
                x1, y1, x2, y2 = box.xyxy[0].cpu().numpy()
                conf = box.conf[0].cpu().numpy()
                cls = int(box.cls[0].cpu().numpy())
                class_name = model.names[cls]
                detections.append(([x1, y1, x2, y2], conf, cls))
    
    # Track objects
    tracked_objects = tracker.update_tracks(detections, frame=frame)
    
    # Reset tracked objects dictionary for this frame
    current_frame_objects = {name: [] for name in model.names.values()}
    
    for track in tracked_objects:
        if not track.is_confirmed():
            continue
        
        track_id = track.track_id
        class_id = track.get_det_class()
        class_name = model.names[class_id]
        x1, y1, x2, y2 = map(int, track.to_tlbr())
        bbox = (x1, y1, x2, y2)
        cx, cy = (x1 + x2) // 2, (y1 + y2) // 2
        
        current_frame_objects[class_name].append(track)
        
        # Draw bounding boxes & IDs
        cv2.rectangle(frame, (x1, y1), (x2, y2), (0, 255, 0), 2)
        cv2.putText(frame, f"{class_name} ID {track_id}", (x1, y1 - 10),
                   cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)

        # -------------------- EVENT DETECTION -------------------- #
        
        # Aircraft Events
        if class_name == "Aircraft":
            if track_id not in aircraft_positions:
                aircraft_positions[track_id] = [(cx, cy)]
            else:
                aircraft_positions[track_id].append((cx, cy))
                if len(aircraft_positions[track_id]) > 30:
                    aircraft_positions[track_id].pop(0)
                
                if len(aircraft_positions[track_id]) >= 30:
                    movement = abs(aircraft_positions[track_id][-1][0] - aircraft_positions[track_id][0][0])
                    detect_aircraft_events(track_id, movement)

        # Passenger Boarding/Deboarding
        if class_name == "Person":
            detect_proximity_events(track_id, bbox, current_frame_objects, 
                                    ["Stairway", "Rolling Stairway"], 
                                    "Passenger Boarding/Deboarding")

        # Baggage Handling
        if class_name == "Baggage Truck":
            detect_proximity_events(track_id, bbox, current_frame_objects, 
                                    ["Aircraft"], "Baggage Handling in Progress", 80)

        # Ramp Loader Activity
        if class_name == "Ramp Loader":
            detect_proximity_events(track_id, bbox, current_frame_objects, 
                                    ["Aircraft"], "Ramp Loader in Action", 80)

        # Bus Arriving
        bus_threshold = frame_width * 0.05  # Dynamic threshold based on video width
        if class_name == "Bus" and x1 < bus_threshold:
            if not event_states.get("bus_arrival", False):
                print("EVENT: Bus Arriving")
                event_states["bus_arrival"] = True

        # Refueling Detection
        if class_name == "Fuel Truck":
            detect_proximity_events(track_id, bbox, current_frame_objects, 
                                    ["Tank Hose"], "Refueling Started")

        # Ground Power Unit Connection
        if class_name == "Ground Power Unit":
            detect_proximity_events(track_id, bbox, current_frame_objects, 
                                    ["Aircraft"], "Ground Power Connected", 80)

    # Update the tracked objects dictionary
    tracked_objects_dict = current_frame_objects
    
    cv2.imshow("Turnaround Tracking", frame)
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

cap.release()
cv2.destroyAllWindows()
```
Would this be a recommended approach for event detection using YOLOv8, or do you suggest any optimizations (e.g., better tracking methods, different post-processing techniques)?

Looking forward to your thoughts. Thanks in advance

**Reply by glenn-jocher** [reply]
Your approach is valid! For optimization, consider:  
1. Using Ultralytics' native [BoT-SORT](https://docs.ultralytics.com/modes/track/#available-trackers) tracker instead of DeepSORT for better integration with YOLOv8 outputs  
2. Implementing multithreading using our [multi-stream tracking example](https://docs.ultralytics.com/modes/track/#how-can-i-run-object-tracking-on-multiple-video-streams-simultaneously) for scalability  

The Ultralytics team appreciates your detailed implementation - great use of spatial-temporal analysis for aviation events!

**Reply by Manishthakur2503** [reply]
@glenn-jocher  First of all, thank you for your response.

I appreciate your suggestions regarding BoT-SORT and multi-stream tracking. I wanted to ask if it would be possible for you to provide some additional help:
- If you already have a reference script for event detection using YOLOv8 and tracking, could you please share it?
- If not, could you guide me on how to detect the boarding process for an aircraft with two stairways (one of class "Stairway" and the other "Rolling Stairway")? 
Specifically, how can I determine when:
- Boarding starts
- Boarding is in progress
- Boarding has ended

If I can implement it for this case, I’ll be able to extend the logic to other event types as well.

Looking forward to your insights. Thanks again for your time and support.

**Reply by glenn-jocher** [reply]
@Manishthakur2503 thank you for your follow-up. While we don't have a dedicated event detection script, you can extend your existing implementation by adding time-threshold checks to your proximity detection. For boarding process monitoring:  

1. Track duration of person-stairway proximity using `time` module timestamps  
2. Implement state transitions with `collections.defaultdict` to track "pre-boarding/boarding/completed" per aircraft  
3. Add frame-count persistence checks to avoid transient detections  

Your current `detect_proximity_events()` function could be extended with duration tracking like this:  
```python
from collections import defaultdict
boarding_timers = defaultdict(lambda: {"start": None, "count": 0})

# Inside your detection loop
if event_name == "Passenger Boarding/Deboarding":
    boarding_timers[track_id]["count"] += 1
    if boarding_timers[track_id]["count"] > FRAME_THRESHOLD: 
        print(f"Boarding in progress for {track_id}")
```

For optimal tracking configuration, see our [BoT-SORT documentation](https://docs.ultralytics.com/modes/track/#available-trackers).
